sdk/resource/resource.go

```diff
@@ -17,101 +17,118 @@
 func New(kvs ...core.KeyValue) *Resource {
-	res := &Resource{keySet: make(map[core.Key]struct{})}
-	for _, kv := range kvs {
-		// First key wins.
-		if _, ok := res.keySet[kv.Key]; !ok {
-			res.keySet[kv.Key] = struct{}{}
-			res.sorted = append(res.sorted, kv)
-		}
+	return &Resource{
+		labels: label.NewSet(kvs...),
 	}
-	sort.Slice(res.sorted, func(i, j int) bool {
-		return res.sorted[i].Key < res.sorted[j].Key
-	})
-	return res
 }
 
-// String implements the Stringer interface and provides a reproducibly
-// hashable representation of a Resource.
-func (r Resource) String() string {
-	// Ensure unique strings if key/value contains '=', ',', or '\'.
-	escaper := strings.NewReplacer("=", `\=`, ",", `\,`, `\`, `\\`)
-
-	var b strings.Builder
-	// Note: this could be further optimized by precomputing the size of
-	// the resulting buffer and adding a call to b.Grow
-	b.WriteString("Resource(")
-	if len(r.sorted) > 0 {
-		b.WriteString(escaper.Replace(string(r.sorted[0].Key)))
-		b.WriteRune('=')
-		b.WriteString(escaper.Replace(r.sorted[0].Value.Emit()))
-		for _, s := range r.sorted[1:] {
-			b.WriteRune(',')
-			b.WriteString(escaper.Replace(string(s.Key)))
-			b.WriteRune('=')
-			b.WriteString(escaper.Replace(s.Value.Emit()))
-		}
-
+// String implements the Stringer interface and provides a
+// human-readable form of the resource.
+//
+// Avoid using this representation as the key in a map of resources,
+// use Equivalent() as the key instead.
+func (r *Resource) String() string {
+	if r == nil {
+		r = Empty()
 	}
-	b.WriteRune(')')
-
-	return b.String()
+	return r.labels.Encoded(label.DefaultEncoder())
 }
 
 // Attributes returns a copy of attributes from the resource in a sorted order.
-func (r Resource) Attributes() []core.KeyValue {
-	return append([]core.KeyValue(nil), r.sorted...)
+// To avoid allocating a new slice, use an iterator.
+func (r *Resource) Attributes() []core.KeyValue {
+	if r == nil {
+		r = Empty()
+	}
+	return r.labels.ToSlice()
 }
 
 // Iter returns an interator of the Resource attributes.
-//
 // This is ideal to use if you do not want a copy of the attributes.
-func (r Resource) Iter() AttributeIterator {
-	return NewAttributeIterator(r.sorted)
+func (r *Resource) Iter() label.Iterator {
+	if r == nil {
+		r = Empty()
+	}
+	return r.labels.Iter()
 }
 
-// Equal returns true if other Resource is equal to r.
-func (r Resource) Equal(other Resource) bool {
-	return r.String() == other.String()
+// Equal returns true when a Resource is equivalent to this Resource.
+func (r *Resource) Equal(eq *Resource) bool {
+	if r == nil {
+		r = Empty()
+	}
+	if eq == nil {
+		eq = Empty()
+	}
+	return r.Equivalent() == eq.Equivalent()
 }
 
 // Merge creates a new resource by combining resource a and b.
 // If there are common key between resource a and b then value from resource a is preserved.
 // If one of the resources is nil then the other resource is returned without creating a new one.
```

Comment:
Don't think this is true anymore, which (it not being true anymore) I also think is appropriate for an immutable Resource.
```suggestion
```

Reply:
Done.